server/middleware/updateActiveAgencies.test.ts

```diff
@@ -0,0 +1,142 @@
+  ]
+  let newActiveAgencies = activeAgenciesResponse.map(agency => agency.agencyId)
+
+  beforeEach(() => {
+    // Mock HMPPS Auth token request **once**
+    fakeAuthServer
+      .post('/oauth/token', 'grant_type=client_credentials')
+      .basicAuth({
+        user: config.apis.hmppsAuth.systemClientId,
+        pass: config.apis.hmppsAuth.systemClientSecret,
+      })
+      .matchHeader('Content-Type', 'application/x-www-form-urlencoded')
+      // Force AuthenticationClient to not cache token
+      .reply(200, { access_token: systemToken, expires_in: 0 })
+
+    activeAgenciesCache.reset()
+  })
+
+  it('updates activeAgencies and applicationInfo singletons', async () => {
+    // Mock Prison API request **once**
+    mockAgenciesSwitchesRequest(activeAgenciesResponse)
+
+    const next: NextFunction = jest.fn()
+
+    // Check initial state
+    expect(activeAgencies).toEqual([])
+    expect(applicationInfo.additionalFields.activeAgencies).toEqual([])
+
+    await middleware(req, res, next)
+
+    // Check activeAgencies was update
+    expect(activeAgencies).toEqual(newActiveAgencies)
+    expect(applicationInfo.additionalFields.activeAgencies).toEqual(newActiveAgencies)
+    expect(next).toHaveBeenNthCalledWith(1)
+  })
+
+  it('uses cache', async () => {
+    // Mock Prison API request **once**
+    mockAgenciesSwitchesRequest(activeAgenciesResponse)
+
+    const next: NextFunction = jest.fn()
+
+    // Recent middleware invocation
+    await middleware(req, res, next)
+
+    // Invoke middleware again to check caching
+    await middleware(req, res, next)
+
+    // NOTE: Here the test is that the nock's Prison API mock (mocked **once**)
+    // didn't receive a 2nd request (that would throw an exception)
+
+    // activeAgencies still the same
+    expect(activeAgencies).toEqual(newActiveAgencies)
+    expect(applicationInfo.additionalFields.activeAgencies).toEqual(newActiveAgencies)
+    expect(next).toHaveBeenNthCalledWith(2)
+  })
+
+  it("converts '*ALL*' to '***' in applicationInfo", async () => {
+    activeAgenciesResponse = [
+      { agencyId: moorland.agencyId, name: moorland.description },
+      { agencyId: brixton.agencyId, name: brixton.description },
+      { agencyId: SERVICE_ALL_AGENCIES, name: 'Service active in all agencies' },
+    ]
+    newActiveAgencies = activeAgenciesResponse.map(agency => agency.agencyId)
+
+    // Mock Prison API request **once**
+    mockAgenciesSwitchesRequest(activeAgenciesResponse)
+
+    const next: NextFunction = jest.fn()
+
+    await middleware(req, res, next)
+
+    // activeAgencies updated to contain only `'*ALL*'`
+    expect(activeAgencies).toEqual([SERVICE_ALL_AGENCIES])
+    // applicationInfo's activeAgencies updated to contain only `'***'`
+    expect(applicationInfo.additionalFields.activeAgencies).toEqual(['***'])
+    expect(next).toHaveBeenNthCalledWith(1)
+  })
+
+  it('handles failure to get a fresh activeAgencies list gracefully', async () => {
```

Comment:
this test depends on the others above it otherwise `activeAgencies` would be empty wouldn’t it? that kinda makes it a bit brittle if so

Reply:
True, I've tweaked that test to set activeAgencies at the start to order shouldn't matter: https://github.com/ministryofjustice/hmpps-incident-reporting/pull/477/commits/2e729df0091398be0284c61b1f74bbd8f8ab8ff1